File: src/main/java/aoc2019/day13/Tile.java
package aoc2019.day13;

import java.util.HashMap;
import java.util.Map;

public enum Tile {
    EMPTY(0, " "),
    WALL(1, "|"),
    BLOCK(2, "■"),
    HORIZONTAL_PADDLE(3, "_"),
    BALL(4, "o");

    private static Map map = new HashMap<>();

    static {
        for (Tile t : Tile.values()) {
            map.put(t.id, t);
        }
    }

    private int id;
    private String symbol;

    Tile(int id, String symbol) {
        this.id = id;
        this.symbol = symbol;
    }

    public static Tile valueOf(int tileId) {
        return (Tile) map.get(tileId);
    }

    public String getSymbol() {
        return symbol;
    }
}

File: src/main/kotlin/aoc2019/day13/Tile.kt
package aoc2019.day13

import java.util.*

enum class Tile(private val id: Int, val symbol: String) {
    EMPTY(0, " "), WALL(1, "|"), BLOCK(2, "■"), HORIZONTAL_PADDLE(3, "_"), BALL(4, "o");

    companion object {
        private val map: MutableMap<Int, Tile> = HashMap<Int, Tile>()

        @JvmStatic
        fun valueOf(tileId: Int): Tile? {
            return map[tileId]
        }

        init {
            for (t in values()) {
                map[t.id] = t
            }
        }
    }

}